### ISSUE TITLE
Use webidl2js named/indexed properties everywhere

### ISSUE THREAD

**Issue by domenic** [jsdom/jsdom#1129]
Now that we have support for named and indexed properties, we should figure out how to use them everywhere---while not regressing performance.

### Existing collections implemented in jsdom, to update

- [x] FileList
- [x] DOMStringMap
- [x] NodeList
- [x] HTMLCollection
- [x] HTMLOptionsCollection
- [x] HTMLSelectElement
- [x] NamedNodeMap #2052
- [x] DOMTokenList #1953
- [ ] HTMLFormElement (currently doesn't do anything)
- [x] Document (aka HTMLDocument in jsdom; currently doesn't do anything)

There's also Window, but that's a whole bunch more work, as it needs a special named properties object, and performance is likely to be even more delicate.

### New collections we can now implement

- Storage, i.e. localStorage and sessionStorage: https://github.com/tmpvar/jsdom/issues/1137

**Comment by Joris-van-der-Wel** [comment]
Some are marked as `[OverrideBuiltins]` http://www.w3.org/TR/WebIDL/#OverrideBuiltins:

* `HTMLFormElement`
* `DOMStringMap`
* `Document`

(but not `HTMLCollection` and `Window`).

This means that the named properties will shadow any property (e.g. toString, attributes, firstChild, \__proto__). This will break a lot. For this to work, jsdom must only use getters/setters/methods that are not exposed (for example, everything protected by symbols).

IMO, this part of the spec is completely insane and I would rather pretend `[OverrideBuiltins]` does not exist. Even if the browsers all implement it this way.

**Comment by Joris-van-der-Wel** [comment]
I think it is okay to do it for `DOMStringMap` though, it has no default properties except for what `Object` provides. So it would be kinda like `Object.create(null)`

**Comment by domenic** [comment]
Yeah, I'd be happy leaving `[OverrideBuiltins]` as a TODO for the far future when none of jsdom goes through the public interface of any other part of jsdom.

**Comment by domenic** [comment]
Renamed in light of https://github.com/tmpvar/jsdom/pull/1934!

**Comment by TimothyGu** [comment]
An update: NodeList and HTMLCollection now use webidl2js. NamedNodeMap (used for element.attributes) *was* moved to webidl2js, but Proxy performance was then found to be too slow for NamedNodeMap. DOMTokenList has been implemented in jsdom, but has not been transitioned yet. It's on my radar though. Additionally, DOMStringMap (used for element.dataset) and HTMLOptionsCollection (a la selectElement.options) have been implemented.

**Comment by domenic** [comment]
Thanks, updated OP further. I misremembered about NodeList.

**Comment by domenic** [comment]
Closing the meta-issue as only https://github.com/jsdom/jsdom/issues/1570 remains and we can track that there.
